erigon-lib/downloader/webseed.go

```diff
@@ -195,3 +204,96 @@ func (d *WebSeeds) readWebSeedsFile(webSeedProviderPath string) (snaptype.WebSee
 }
+
+// downloadTorrentFilesFromProviders - if they are not exist on file-system
+func (d *WebSeeds) downloadTorrentFilesFromProviders(ctx context.Context, rootDir string) {
+	// TODO: need more tests, need handle more forward-compatibility and backward-compatibility case
+	//  - now, if add new type of .torrent files to S3 bucket - existing nodes will start downloading it. maybe need whitelist of file types
+	//  - maybe need download new files if --snap.stop=true
+	if !d.downloadTorrentFile {
+		return
+	}
+	if len(d.TorrentUrls()) == 0 {
+		return
+	}
+	var addedNew int
+	e, ctx := errgroup.WithContext(ctx)
+	urlsByName := d.TorrentUrls()
+	//TODO:
+	// - what to do if node already synced?
+	for name, tUrls := range urlsByName {
+		tPath := filepath.Join(rootDir, name)
+		if dir.FileExist(tPath) {
+			continue
+		}
+		addedNew++
+		if strings.HasSuffix(name, ".v.torrent") || strings.HasSuffix(name, ".ef.torrent") {
+			_, fName := filepath.Split(name)
+			if strings.HasPrefix(fName, "commitment") {
+				d.logger.Log(d.verbosity, "[snapshots] webseed has .torrent, but we skip it because we don't support it yet", "name", name)
+				continue
+			}
+		}
+		name := name
+		tUrls := tUrls
+		e.Go(func() error {
+			for _, url := range tUrls {
+				res, err := d.callTorrentHttpProvider(ctx, url)
+				if err != nil {
+					d.logger.Debug("[snapshots] callTorrentHttpProvider", "err", err)
+					continue
+				}
+				d.logger.Log(d.verbosity, "[snapshots] downloaded .torrent file from webseed", "name", name)
+				if err := saveTorrent(tPath, res); err != nil {
+					d.logger.Debug("[snapshots] saveTorrent", "err", err)
+					continue
+				}
+				return nil
+			}
+			return nil
+		})
+	}
+	if err := e.Wait(); err != nil {
+		d.logger.Debug("[snapshots] webseed discover", "err", err)
+	}
+}
+
+func (d *WebSeeds) callTorrentHttpProvider(ctx context.Context, url *url.URL) ([]byte, error) {
+	request, err := http.NewRequest(http.MethodGet, url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	request = request.WithContext(ctx)
+	resp, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return nil, fmt.Errorf("webseed.downloadTorrentFile: host=%s, url=%s, %w", url.Hostname(), url.EscapedPath(), err)
+	}
+	defer resp.Body.Close()
+	//protect against too small and too big data
+	if resp.ContentLength == 0 || resp.ContentLength > int64(128*datasize.MB) {
+		return nil, nil
+	}
+	res, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("webseed.downloadTorrentFile: host=%s, url=%s, %w", url.Hostname(), url.EscapedPath(), err)
+	}
+	if err = validateTorrentBytes(res, d.torrentHashes); err != nil {
+		return nil, fmt.Errorf("webseed.downloadTorrentFile: host=%s, url=%s, %w", url.Hostname(), url.EscapedPath(), err)
+	}
+	return res, nil
+}
+
+func validateTorrentBytes(b []byte, torrentHashes []string) error {
+	var mi metainfo.MetaInfo
+	if len(torrentHashes) == 0 {
+		return fmt.Errorf("no snapshots provided")
```

Comment:
isn't 0 hashes - is valid case? for example if you run new chain - which has 0 snapshots yet.